File: python/08_001.py
def has_same_pattern(str1, str2):
  if len(str1) != len(str2):
     return False
  
  map1 = {}
  map2 = {}
  
  for c1, c2 in zip(str1, str2):
     if c1 in map1:
        if map1[c1] != c2:
           return False
     else:
        map1[c1] = c2
     
     if c2 in map2:
        if map2[c2] != c1:
           return False
     else:
        map2[c2] = c1
  
  return True

File: java/08_001.java
import java.util.HashMap;
import java.util.Map;

public boolean hasSamePattern(String str1, String str2) {
    if (str1.length() != str2.length()) {
        return false;
    }
    
    Map<Character, Character> map1 = new HashMap<>();
    Map<Character, Character> map2 = new HashMap<>();
    
    for (int i = 0; i < str1.length(); i++) {
        char c1 = str1.charAt(i);
        char c2 = str2.charAt(i);
        
        // Check mapping from str1 to str2
        if (map1.containsKey(c1)) {
            if (map1.get(c1) != c2) {
                return false;
            }
        } else {
            map1.put(c1, c2);
        }
        
        // Check mapping from str2 to str1
        if (map2.containsKey(c2)) {
            if (map2.get(c2) != c1) {
                return false;
            }
        } else {
            map2.put(c2, c1);
        }
    }
    
    return true;
}